lib/url-decorator.jsx

```diff
@@ -0,0 +1,53 @@
+import React from 'react';
+import { CompositeDecorator } from 'draft-js';
+import urlRegex from 'url-regex';
+import normalizeUrl from 'normalize-url';
+import { viewExternalUrl } from './utils/url-utils';
+
+const urlRe = urlRegex();
+/*
+- markdownLinkRe is a regexp to be applied to the substring preceding a matched url, ie. `wholeString.slice(0, wholeString.indexOf(matchedUrl))`
+- it tests whether the matched url is preceded by the charcters that define a link in markdown, ie. `(` or `]: `
+*/
+const markdownLinkRe = /(?:\(|\]:\s+)$/;
+
+export function normalizeUrlsInMarkdownString( str ) {
+	return str.replace( urlRe, ( match, offset ) => {
+		var normalized = normalizeUrl( match );
+		if ( markdownLinkRe.test( str.slice( 0, offset ) ) ) {
+			return normalized;
+		}
+		// replace url with markdown link only if it's not already part of one
+		return `[${match}](${normalized})`;
+	} );
+}
```

Comment:
This function could get really expensive as we add URLs and as the document length increases. At a minimum I believe that we could rearrange it to scan through the text instead of iterating over the entire document-up-until-the-start-of-the-URL.

Maybe through the use of splitting on the RegExp matches we can build a recursive parser that would eliminate the overhead of starting fresh each time. It's something I'll think about more and follow-up here, but feel free to add your own ideas.

I do not believe that the performance of this function as written will cut it, unfortunately 😢 but we will prevail!